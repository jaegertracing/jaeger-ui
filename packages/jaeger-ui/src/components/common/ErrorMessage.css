/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.ErrorMessage {
  position: relative;
  word-break: break-word;
  word-wrap: break-word;
}

.ErrorMessage--msg {
  color: #c00;
  margin: 0;
  padding-bottom: 1rem;
}

.ErrorMessage--details {
  background: #f5f5f5;
  border: 1px solid #e6e6e6;
  overflow: auto;
}

.ErrorMessage--detailsTable {
  min-width: 100%;
}

.ErrorMessage--detailItem:nth-child(2n) > .ErrorMessage--attr,
.ErrorMessage--detailItem:nth-child(2n) > .ErrorMessage--value {
  background-color: #f0f0f0;
  border-bottom: 1px solid #e8e8e8;
  border-top: 1px solid #e8e8e8;
}

/* dont have extra borders on the bottom row */
.ErrorMessage--detailItem:last-child > .ErrorMessage--attr,
.ErrorMessage--detailItem:last-child > .ErrorMessage--value {
  border-bottom: none;
}

.ErrorMessage--attr {
  color: rgba(0, 0, 0, 0.45);
  padding: 0.5em;
  vertical-align: top;
  white-space: nowrap;
}

.ErrorMessage--value {
  color: rgba(0, 0, 0, 0.8);
  font-family: monospace;
  padding: 0.5em;
  white-space: pre;
}
