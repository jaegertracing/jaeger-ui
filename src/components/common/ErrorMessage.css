/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.ErrorMessage {
  background: #f5f5f5;
  border: 1px solid #e6e6e6;
  position: relative;
  word-break: break-word;
  word-wrap: break-word;
}

.ErrorMessage--msg {
  color: #c00;
  margin: 0;
  padding: 1rem;
}

.ErrorMessage--details {
  overflow: auto;
  padding: 0 1rem 1rem 1rem;
}

.ErrorMessage--attr {
  padding-bottom: 0.5em;
  padding-right: 1em;
  vertical-align: top;
  white-space: nowrap;
}

.ErrorMessage--value {
  font-family: monospace;
  white-space: pre;
}
