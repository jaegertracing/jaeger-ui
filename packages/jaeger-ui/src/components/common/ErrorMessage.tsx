// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { ApiError } from '../../types/api-error';

import './ErrorMessage.css';

type ErrorMessageProps = {
  className?: string;
  detailClassName?: string;
  messageClassName?: string;
  error: ApiError;
};

type SubPartProps = {
  className?: string;
  error: ApiError;
  wrap?: boolean;
  wrapperClassName?: string;
};

type ErrorAttrProps = {
  name: string;
  value: string | number;
};

export const MAX_DETAIL_LENGTH = 1024;

function ErrorAttr({ name, value }: ErrorAttrProps) {
  return (
    <tr className="ErrorMessage--detailItem">
      <td className="ErrorMessage--attr">{name}</td>
      <td className="ErrorMessage--value">{value}</td>
    </tr>
  );
}

export function Message(props: SubPartProps) {
  const { className, error, wrap, wrapperClassName } = props;
  const cssClass = `ErrorMessage--msg ${className || ''}`;

  const msg =
    typeof error === 'string' ? (
      <h3 className={cssClass}>{error}</h3>
    ) : (
      <h3 className={cssClass}>{error.message}</h3>
    );

  if (wrap) {
    return <div className={`ErrorMessage ${wrapperClassName || ''}`}>{msg}</div>;
  }

  return msg;
}

Message.defaultProps = {
  className: undefined,
  wrap: false,
  wrapperClassName: undefined,
};

export function Details(props: SubPartProps) {
  const { className, error, wrap, wrapperClassName } = props;

  if (typeof error === 'string') {
    return null;
  }

  const { httpStatus, httpStatusText, httpUrl, httpQuery, httpBody } = error;
  const bodyExcerpt =
    httpBody && httpBody.length > MAX_DETAIL_LENGTH
      ? `${httpBody.slice(0, MAX_DETAIL_LENGTH - 3).trim()}...`
      : httpBody;

  const details = (
    <div
      className={`ErrorMessage--details ${className || ''} u-simple-scrollbars`}
      data-testid="ErrorMessage--details"
    >
      <table className="ErrorMessage--detailsTable">
        <tbody>
          {httpStatus && <ErrorAttr name="Status" value={httpStatus} />}
          {httpStatusText && <ErrorAttr name="Status text" value={httpStatusText} />}
          {httpUrl && <ErrorAttr name="URL" value={httpUrl} />}
          {httpQuery && <ErrorAttr name="Query" value={httpQuery} />}
          {bodyExcerpt && <ErrorAttr name="Response body" value={bodyExcerpt} />}
        </tbody>
      </table>
    </div>
  );

  if (wrap) {
    return (
      <div className={`ErrorMessage ${wrapperClassName || ''}`} data-testid="ErrorMessage--details--wrapper">
        {details}
      </div>
    );
  }

  return details;
}

Details.defaultProps = {
  className: undefined,
  wrap: false,
  wrapperClassName: undefined,
};

export default function ErrorMessage({
  className,
  detailClassName,
  error,
  messageClassName,
}: ErrorMessageProps) {
  if (!error) {
    return null;
  }

  if (typeof error === 'string') {
    return <Message className={messageClassName} error={error} wrapperClassName={className} wrap />;
  }

  return (
    <div className={`ErrorMessage ${className || ''}`} data-testid="ErrorMessage">
      <Message error={error} className={messageClassName} />
      <Details error={error} className={detailClassName} />
    </div>
  );
}

ErrorMessage.defaultProps = {
  className: undefined,
  detailClassName: undefined,
  messageClassName: undefined,
};
