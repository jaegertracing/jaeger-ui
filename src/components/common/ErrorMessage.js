// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';

import './ErrorMessage.css';

type ErrorMessageProps = {
  className?: string,
  error:
    | string
    | {
        message: string,
        httpStatus?: any,
        httpStatusText?: string,
        httpUrl?: string,
        httpQuery?: string,
        httpBody?: string,
      },
};

function ErrorAttr({ name, value }: { name: string, value: any }) {
  return (
    <tr>
      <td className="ErrorMessage--attr">{name}</td>
      <td className="ErrorMessage--value">{value}</td>
    </tr>
  );
}

export default function ErrorMessage({ className, error }: ErrorMessageProps) {
  if (!error) {
    return null;
  }
  if (typeof error === 'string') {
    return (
      <div className={`ErrorMessage ${className || ''}`}>
        <h3 className="ErrorMessage--msg">{error}</h3>
      </div>
    );
  }
  const { message, httpStatus, httpStatusText, httpUrl, httpQuery, httpBody } = error;
  const bodyExcerpt = httpBody && httpBody.length > 1024 ? `${httpBody.slice(0, 1021).trim()}...` : httpBody;
  return (
    <div className={`ErrorMessage ${className || ''}`}>
      <h2 className="ErrorMessage--msg">{message}</h2>
      <div className="ErrorMessage--details u-simple-scrollbars">
        <table>
          <tbody>
            {httpStatus ? <ErrorAttr name="Status" value={httpStatus} /> : null}
            {httpStatusText ? <ErrorAttr name="Status text" value={httpStatusText} /> : null}
            {httpUrl ? <ErrorAttr name="URL" value={httpUrl} /> : null}
            {httpQuery ? <ErrorAttr name="Query" value={httpQuery} /> : null}
            {bodyExcerpt ? <ErrorAttr name="Response body" value={bodyExcerpt} /> : null}
          </tbody>
        </table>
      </div>
    </div>
  );
}

ErrorMessage.defaultProps = {
  className: undefined,
};
