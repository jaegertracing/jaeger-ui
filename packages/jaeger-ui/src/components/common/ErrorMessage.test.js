// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow, mount } from 'enzyme';

import ErrorMessage from './ErrorMessage';

describe('<ErrorMessage>', () => {
  let wrapper;
  let error;

  beforeEach(() => {
    error = 'some-error';
    wrapper = shallow(<ErrorMessage error={error} />);
  });

  it('is ok when not passed an error', () => {
    wrapper.setProps({ error: null });
    expect(wrapper).toBeDefined();
  });

  it('renders a message when passed a string', () => {
    const msg = wrapper.find('Message');
    expect(msg.length).toBe(1);
    expect(msg.shallow().text()).toMatch(error);
  });

  it('<Details /> renders empty on string error', () => {
    wrapper = shallow(<ErrorMessage.Details error={error} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('<Details /> renders wrapper on wrap', () => {
    error = {
      message: 'some-http-ish-message',
      httpStatusText: 'value-httpStatusText',
      httpUrl: 'value-httpUrl',
      httpQuery: 'value-httpQuery',
      httpBody: 'value-httpBody',
    };
    wrapper = shallow(<ErrorMessage.Details error={error} wrap />);
    expect(wrapper).toMatchSnapshot();
  });

  it('<Details /> renders custom wrapper class', () => {
    error = {
      message: 'some-http-ish-message',
      httpStatusText: 'value-httpStatusText',
      httpUrl: 'value-httpUrl',
      httpQuery: 'value-httpQuery',
      httpBody: 'value-httpBody',
    };
    wrapper = shallow(<ErrorMessage.Details error={error} wrap wrapperClassName="TEST-WRAPPER-CLASS" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('rendering more complex errors', () => {
    it('renders the error message', () => {
      error = new Error('another-error');
      wrapper.setProps({ error });
      const msg = wrapper.find('Message');
      expect(msg.length).toBe(1);
      expect(msg.shallow().text()).toMatch(error.message);
    });

    it('renders HTTP related data from the error', () => {
      error = {
        message: 'some-http-ish-message',
        httpStatus: 'value-httpStatus',
        httpStatusText: 'value-httpStatusText',
        httpUrl: 'value-httpUrl',
        httpQuery: 'value-httpQuery',
        httpBody: 'value-httpBody',
      };
      wrapper.setProps({ error });
      const details = wrapper.find('Details');
      expect(details.length).toBe(1);
      const detailsWrapper = details.shallow();
      Object.keys(error).forEach(key => {
        if (key === 'message') {
          return;
        }
        const errorAttr = detailsWrapper.find(`ErrorAttr[value="${error[key]}"]`);
        expect(errorAttr.length).toBe(1);
      });
    });

    it('renders truncated body excerpt on large body', () => {
      error = {
        message: 'some-http-ish-message',
        httpStatus: 'value-httpStatus',
        httpStatusText: 'value-httpStatusText',
        httpUrl: 'value-httpUrl',
        httpQuery: 'value-httpQuery',
        httpBody: `value-httpBody-Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vel
          arcu mattis, pellentesque metus eget, fermentum dolor. Pellentesque habitant morbi
          tristique senectus et netus et malesuada fames ac turpis egestas. Integer placerat turpis
          diam, in eleifend est congue maximus. Sed dictum ipsum a dolor pellentesque, vestibulum
          congue mi pharetra. Pellentesque id augue sit amet libero consectetur rutrum ac rutrum
          nisl. Proin suscipit, libero et gravida tristique, magna dolor pulvinar ex, sit amet
          accumsan tellus metus vel elit. Ut condimentum ultrices felis vel feugiat. Praesent tellus
          nisl, feugiat ut est sed, lobortis lacinia elit. Aenean consequat nunc arcu, in tincidunt
          libero blandit non.
          
          Vestibulum aliquet lobortis tincidunt. Donec in nunc a sapien rutrum ultricies eu ac
          felis. Vestibulum feugiat in mauris id fringilla. Nulla enim nisl, rutrum vel tellus vel,
          tincidunt aliquam enim. Nullam in tortor mi. Vestibulum pellentesque velit eget eros
          varius, ut sollicitudin urna consectetur. Ut gravida, dolor id varius faucibus, magna
          lorem condimentum metus, sit amet hendrerit nisi nisi eu dolor. Nulla vel metus sit amet
          odio rhoncus rutrum eu in sapien. Mauris aliquam sed sem sit amet dignissim.`,
      };

      wrapper.setProps({ error });
      const details = wrapper.find('Details');
      expect(details.length).toBe(1);
      const detailsWrapper = details.shallow();
      expect(detailsWrapper).toMatchSnapshot();
    });

    it('renders on missing httpStatus from the error', () => {
      error = {
        message: 'some-http-ish-message',
        httpStatusText: 'value-httpStatusText',
        httpUrl: 'value-httpUrl',
        httpQuery: 'value-httpQuery',
        httpBody: 'value-httpBody',
      };
      wrapper.setProps({ error });
      const details = wrapper.find('Details');
      expect(details.length).toBe(1);
      const detailsWrapper = details.shallow();
      expect(detailsWrapper).toMatchSnapshot();
    });
  });

  it('is fine when mounted', () => {
    error = { message: 'le-error', httpStatus: 'some-status' };
    wrapper = mount(<ErrorMessage error={error} />);
    expect(wrapper).toBeDefined();
  });
});
