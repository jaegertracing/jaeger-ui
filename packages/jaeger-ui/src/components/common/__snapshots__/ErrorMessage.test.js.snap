// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorMessage> <Details /> renders custom wrapper class 1`] = `
<div
  className="ErrorMessage TEST-WRAPPER-CLASS"
>
  <div
    className="ErrorMessage--details  u-simple-scrollbars"
  >
    <table
      className="ErrorMessage--detailsTable"
    >
      <tbody>
        <ErrorAttr
          name="Status text"
          value="value-httpStatusText"
        />
        <ErrorAttr
          name="URL"
          value="value-httpUrl"
        />
        <ErrorAttr
          name="Query"
          value="value-httpQuery"
        />
        <ErrorAttr
          name="Response body"
          value="value-httpBody"
        />
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<ErrorMessage> <Details /> renders empty on string error 1`] = `""`;

exports[`<ErrorMessage> <Details /> renders wrapper on wrap 1`] = `
<div
  className="ErrorMessage "
>
  <div
    className="ErrorMessage--details  u-simple-scrollbars"
  >
    <table
      className="ErrorMessage--detailsTable"
    >
      <tbody>
        <ErrorAttr
          name="Status text"
          value="value-httpStatusText"
        />
        <ErrorAttr
          name="URL"
          value="value-httpUrl"
        />
        <ErrorAttr
          name="Query"
          value="value-httpQuery"
        />
        <ErrorAttr
          name="Response body"
          value="value-httpBody"
        />
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<ErrorMessage> rendering more complex errors renders on missing httpStatus from the error 1`] = `
<div
  className="ErrorMessage--details  u-simple-scrollbars"
>
  <table
    className="ErrorMessage--detailsTable"
  >
    <tbody>
      <ErrorAttr
        name="Status text"
        value="value-httpStatusText"
      />
      <ErrorAttr
        name="URL"
        value="value-httpUrl"
      />
      <ErrorAttr
        name="Query"
        value="value-httpQuery"
      />
      <ErrorAttr
        name="Response body"
        value="value-httpBody"
      />
    </tbody>
  </table>
</div>
`;

exports[`<ErrorMessage> rendering more complex errors renders truncated body excerpt on large body 1`] = `
<div
  className="ErrorMessage--details  u-simple-scrollbars"
>
  <table
    className="ErrorMessage--detailsTable"
  >
    <tbody>
      <ErrorAttr
        name="Status"
        value="value-httpStatus"
      />
      <ErrorAttr
        name="Status text"
        value="value-httpStatusText"
      />
      <ErrorAttr
        name="URL"
        value="value-httpUrl"
      />
      <ErrorAttr
        name="Query"
        value="value-httpQuery"
      />
      <ErrorAttr
        name="Response body"
        value="value-httpBody-Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vel
          arcu mattis, pellentesque metus eget, fermentum dolor. Pellentesque habitant morbi
          tristique senectus et netus et malesuada fames ac turpis egestas. Integer placerat turpis
          diam, in eleifend est congue maximus. Sed dictum ipsum a dolor pellentesque, vestibulum
          congue mi pharetra. Pellentesque id augue sit amet libero consectetur rutrum ac rutrum
          nisl. Proin suscipit, libero et gravida tristique, magna dolor pulvinar ex, sit amet
          accumsan tellus metus vel elit. Ut condimentum ultrices felis vel feugiat. Praesent tellus
          nisl, feugiat ut est sed, lobortis lacinia elit. Aenean consequat nunc arcu, in tincidunt
          libero blandit non.
          
          Vestibulum aliquet lobortis tincidunt. Donec in nunc a sapien rutrum ultricies eu ac
          felis. Vestibulum feugiat in mauris id fringilla. Nulla enim nisl, rutrum vel tellus vel,
          tincidunt a..."
      />
    </tbody>
  </table>
</div>
`;
